Stutter on Raspberry Pi 400 using example

NOTE: I'm actively investigating this, but I thought I should put in an issue in case others are (or aren't) seeing the same problem.

Raspberry Pi 400 (ARM v8)
Default Pi OS that ships with Pi 400, but updated via apt to latest.
HDMI audio.
go 1.15.4

Effects:

* `go run example/main.go` using master at time of posting ( 7f7e156100089346d897740cd0eba0a00c488751 ). The notes are not clear, they stutter quickly and continuously at the same interval.
  * If I increase `bufferSizeInBytes` in the `NewContext` call the stutter becomes less frequent, when I reach 80KB (the example starts at 4KB) it is an occasional pop every few seconds.
  * CPU usage is <10% on all cores when testing.
* When using Oto via Ebiten to play mp3 music it plays slowly and with stutter - not sure if that's indicative of buffer under-run (maybe specific to HDMI audio?)?

`aplay` plays wavs on the Pi without issue.

```
~/code/src/github.com/TheMightyGit $ git clone git@github.com:TheMightyGit/oto.git
Cloning into 'oto'...
remote: Enumerating objects: 90, done.
remote: Counting objects: 100% (90/90), done.
remote: Compressing objects: 100% (66/66), done.
remote: Total 822 (delta 55), reused 55 (delta 24), pack-reused 732
Receiving objects: 100% (822/822), 185.47 KiB | 1.13 MiB/s, done.
Resolving deltas: 100% (518/518), done.
~/code/src/github.com/TheMightyGit $ cd oto/
~/code/src/github.com/TheMightyGit/oto $ git checkout debug-buffer-sizes
Branch 'debug-buffer-sizes' set up to track remote branch 'debug-buffer-sizes' from 'origin'.
Switched to a new branch 'debug-buffer-sizes'
~/code/src/github.com/TheMightyGit/oto $ go run example/main.go
# github.com/hajimehoshi/oto
./driver_linux.go: In function ‘ALSA_hw_params’:
./driver_linux.go:51:9: warning: format ‘%d’ expects argument of type ‘int’, but argument 2 has type ‘snd_pcm_uframes_t {aka long unsigned int}’ [-Wformat=]
  printf("buffer_size %d\n", *buffer_size);
         ^
./driver_linux.go:53:9: warning: format ‘%d’ expects argument of type ‘int’, but argument 2 has type ‘snd_pcm_uframes_t {aka long unsigned int}’ [-Wformat=]
  printf("near buffer_size %d\n", *buffer_size);
         ^
./driver_linux.go:54:9: warning: format ‘%d’ expects argument of type ‘int’, but argument 2 has type ‘snd_pcm_uframes_t {aka long unsigned int}’ [-Wformat=]
  printf("period_size %d\n", *period_size);
         ^
./driver_linux.go:56:9: warning: format ‘%d’ expects argument of type ‘int’, but argument 2 has type ‘snd_pcm_uframes_t {aka long unsigned int}’ [-Wformat=]
  printf("near period_size %d\n", *period_size);
         ^
2020/11/22 08:09:23 Before ALSA_hw_params: 1024 1024
buffer_size 1024
near buffer_size 1024
period_size 1024
near period_size 341
2020/11/22 08:09:23 After ALSA_hw_params: 1024 341
~/code/src/github.com/TheMightyGit/oto $ 
```
```
~/code/src/github.com/TheMightyGit/oto $ git checkout rpi400-fix
Branch 'rpi400-fix' set up to track remote branch 'rpi400-fix' from 'origin'.
Switched to a new branch 'rpi400-fix'
~/code/src/github.com/TheMightyGit/oto $ go run example/main.go
# github.com/hajimehoshi/oto
./driver_linux.go: In function ‘ALSA_hw_params’:
./driver_linux.go:51:9: warning: format ‘%d’ expects argument of type ‘int’, but argument 2 has type ‘snd_pcm_uframes_t {aka long unsigned int}’ [-Wformat=]
  printf("buffer_size %d\n", *buffer_size);
         ^
./driver_linux.go:53:9: warning: format ‘%d’ expects argument of type ‘int’, but argument 2 has type ‘snd_pcm_uframes_t {aka long unsigned int}’ [-Wformat=]
  printf("near buffer_size %d\n", *buffer_size);
         ^
./driver_linux.go:54:9: warning: format ‘%d’ expects argument of type ‘int’, but argument 2 has type ‘snd_pcm_uframes_t {aka long unsigned int}’ [-Wformat=]
  printf("period_size %d\n", *period_size);
         ^
./driver_linux.go:56:9: warning: format ‘%d’ expects argument of type ‘int’, but argument 2 has type ‘snd_pcm_uframes_t {aka long unsigned int}’ [-Wformat=]
  printf("near period_size %d\n", *period_size);
         ^
2020/11/22 08:10:13 Before ALSA_hw_params: 2048 1024
buffer_size 2048
near buffer_size 2048
period_size 1024
near period_size 682
2020/11/22 08:10:13 After ALSA_hw_params: 2048 682
~/code/src/github.com/TheMightyGit/oto $ 
```

Sounds the same.  There's a click as each note disappears, but I would assume that's due to a lack of zero crossing cutting.

I notice I'm getting a bunch of warnings that you're not.  Looks like a cast is missing somewhere perhaps.

--- a/driver_linux.go
+++ b/driver_linux.go
@@ -118,6 +118,21 @@ func newDriver(sampleRate, numChans, bitDepthInBytes, bufferSizeInBytes int) (tr
 		return nil, alsaError(errCode)
 	}
 
+	// Raspberry Pi 400 fix - for some reason the call to ALSA_hw_params can result in the bufferSize and periodSize
+	// being the same (this could be becuase the rpi driver does not drop the periodSize below 1024, whereas desktop
+	// Linux is happy to go lower). When the bufferSize and periodSize are the same it causes constant buffer underruns
+	// that stutter the audio continuously.
+	//
+	// So in the case where an RPi (or any other driver) returns the same buffer and period size we will change the
+	// buffer size to twice that of the period size and re-request hw params.
+	if bufferSize == periodSize {
+		bufferSize = periodSize * 2
+		if errCode := C.ALSA_hw_params(p.handle, C.uint(sampleRate), C.uint(numChans), format, &bufferSize, &periodSize); errCode < 0 {
+			p.Close()
+			return nil, alsaError(errCode)
+		}
+	}
+
 	// allocate the buffer of the size of the period, use the periodSize that we've got back
 	// from ALSA after it's wise decision
 	p.bufSamples = int(periodSize)

